File: sdk-python/obsvr/strict_receipt.py
"""Pure canonical strict receipts and offline verification."""

from __future__ import annotations

import base64
import binascii
import hashlib
from typing import Any, Dict, List, Optional

from .aarm_outcome import AARM_OUTCOMES
from .tool_pinning import _canonical_json_for_hash

STRICT_RECEIPT_SCHEMA = "obsvr-strict-receipt-v1"
STRICT_RECEIPT_PROFILE_VERSION = "1.0"
STRICT_RECEIPT_ENVELOPE_SCHEMA = "obsvr-strict-receipt-envelope-v1"
STRICT_RECEIPT_BODY_DOMAIN = b"obsvr-strict-receipt/body/1"
STRICT_RECEIPT_SIGNATURE_DOMAIN = b"obsvr-strict-receipt/signature/1"

_OUTCOMES = frozenset(AARM_OUTCOMES)
_HEX = frozenset("0123456789abcdef")
_MAX_SAFE_INTEGER = 9_007_199_254_740_991


class StrictReceiptValidationError(ValueError):
    """The value cannot be represented as one strict receipt."""


def _record(value: Any, field: str) -> Dict[str, Any]:
    if not isinstance(value, dict):
        raise StrictReceiptValidationError(f"{field} must be an object")
    return value


def _exact(value: Dict[str, Any], allowed: set[str], field: str) -> None:
    unknown = sorted(set(value) - allowed)
    if unknown:
        raise StrictReceiptValidationError(
            f"{field} contains unsupported field: {unknown[0]}"
        )


def _text(value: Any, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise StrictReceiptValidationError(f"{field} must be a nonblank string")
    return value


def _hash(value: Any, field: str) -> str:
    if (
        not isinstance(value, str)
        or len(value) != 64
        or any(char not in _HEX for char in value)
    ):
        raise StrictReceiptValidationError(
            f"{field} must be 64 lowercase hex characters"
        )
    return value


def _safe_integer(value: Any, field: str, *, positive: bool = False) -> int:
    floor = 1 if positive else 0
    if (
        isinstance(value, bool)
        or not isinstance(value, int)
        or value < floor
        or value > _MAX_SAFE_INTEGER
    ):
        label = "positive" if positive else "nonnegative"
        raise StrictReceiptValidationError(f"{field} must be a {label} safe integer")
    return value


def _optional_text(source: Dict[str, Any], key: str, field: str) -> Optional[str]:
    if key not in source:
        return None
    return _text(source[key], field)


def _string_set(value: Any, field: str) -> List[str]:
    if not isinstance(value, list):
        raise StrictReceiptValidationError(f"{field} must be an array")
    values = [_text(item, f"{field}[{index}]") for index, item in enumerate(value)]
    return sorted(set(values), key=lambda item: tuple(ord(char) for char in item))


def _u64(value: int) -> bytes:
    return value.to_bytes(8, "big", signed=False)


def strict_receipt_key_id(raw_public_key: bytes) -> str:
    if len(raw_public_key) != 32:
        raise StrictReceiptValidationError("public key must be 32 raw bytes")
    return "sha256:" + hashlib.sha256(raw_public_key).hexdigest()


def build_strict_receipt_body(input_value: Any) -> Dict[str, Any]:
    root = _record(input_value, "receipt body")
    _exact(
        root,
        {
            "schema",
            "profile_version",
            "record_type",
            "receipt_id",
            "session_id",
            "sequence",
            "timestamp_ms",
            "clock_regression_clamped",
            "previous_receipt_hash",
            "sdk",
            "initiator",
            "action",
            "context",
            "evaluation",
            "execution_authorized",
            "suspension",
            "resolution",
        },
        "receipt body",
    )
    if root.get("schema") != STRICT_RECEIPT_SCHEMA:
        raise StrictReceiptValidationError("invalid receipt schema")
    if root.get("profile_version") != STRICT_RECEIPT_PROFILE_VERSION:
        raise StrictReceiptValidationError("invalid receipt profile_version")
    record_type = root.get("record_type")
    if record_type not in ("decision", "resolution"):
        raise StrictReceiptValidationError("invalid record_type")
    session_id = _text(root.get("session_id"), "session_id")
    sequence = _safe_integer(root.get("sequence"), "sequence", positive=True)
    if root.get("receipt_id") != f"{session_id}:{sequence}":
        raise StrictReceiptValidationError("receipt_id must equal session_id:sequence")
    timestamp = _safe_integer(root.get("timestamp_ms"), "timestamp_ms")
    if not isinstance(root.get("clock_regression_clamped"), bool):
        raise StrictReceiptValidationError(
            "clock_regression_clamped must be a boolean"
        )
    if sequence == 1:
        if root.get("previous_receipt_hash") is not None:
            raise StrictReceiptValidationError(
                "genesis previous_receipt_hash must be null"
            )
        previous = None
    else:
        previous = _hash(root.get("previous_receipt_hash"), "previous_receipt_hash")

    sdk = _record(root.get("sdk"), "sdk")
    _exact(sdk, {"language", "version"}, "sdk")
    if sdk.get("language") not in ("typescript", "python"):
        raise StrictReceiptValidationError("invalid sdk.language")
    initiator = _record(root.get("initiator"), "initiator")
    _exact(initiator, {"agent_id", "key_id"}, "initiator")
    key_id = _text(initiator.get("key_id"), "initiator.key_id")
    if (
        not key_id.startswith("sha256:")
        or len(key_id) != 71
        or any(char not in _HEX for char in key_id[7:])
    ):
        raise StrictReceiptValidationError(
            "initiator.key_id must be sha256:<64 lowercase hex>"
        )

    action = _record(root.get("action"), "action")
    _exact(
        action,
        {
            "action_id",
            "kind",
            "name",
            "arguments_hash",
            "target",
            "effective_arguments_hash",
        },
        "action",
    )
    normalized_action: Dict[str, Any] = {
        "action_id": _text(action.get("action_id"), "action.action_id"),
        "kind": _text(action.get("kind"), "action.kind"),
        "name": _text(action.get("name"), "action.name"),
        "arguments_hash": _hash(action.get("arguments_hash"), "action.arguments_hash"),
    }
    target = _optional_text(action, "target", "action.target")
    if target is not None:
        normalized_action["target"] = target

    context = _record(root.get("context"), "context")
    _exact(context, {"schema", "context_hash", "run_id", "thread_id"}, "context")
    if context.get("schema") != "obsvr-action-context-v1":
        raise StrictReceiptValidationError("invalid context.schema")
    normalized_context: Dict[str, Any] = {
        "schema": "obsvr-action-context-v1",
        "context_hash": _hash(context.get("context_hash"), "context.context_hash"),
        "run_id": _text(context.get("run_id"), "context.run_id"),
    }
    thread_id = _optional_text(context, "thread_id", "context.thread_id")
    if thread_id is not None:
        normalized_context["thread_id"] = thread_id

    evaluation = _record(root.get("evaluation"), "evaluation")
    _exact(
        evaluation,
        {
            "input_hash",
            "policy_hash",
            "evaluator_hash",
            "engine_version",
            "policy_version",
            "outcome",
            "reason_code",
            "rule_ids",
        },
        "evaluation",
    )
    outcome = evaluation.get("outcome")
    if not isinstance(outcome, str) or outcome not in _OUTCOMES:
        raise StrictReceiptValidationError("invalid evaluation.outcome")
    normalized_evaluation = {
        "input_hash": _hash(evaluation.get("input_hash"), "evaluation.input_hash"),
        "policy_hash": _hash(evaluation.get("policy_hash"), "evaluation.policy_hash"),
        "evaluator_hash": _hash(
            evaluation.get("evaluator_hash"), "evaluation.evaluator_hash"
        ),
        "engine_version": _text(
            evaluation.get("engine_version"), "evaluation.engine_version"
        ),
        "policy_version": _text(
            evaluation.get("policy_version"), "evaluation.policy_version"
        ),
        "outcome": outcome,
        "reason_code": _text(evaluation.get("reason_code"), "evaluation.reason_code"),
        "rule_ids": _string_set(evaluation.get("rule_ids"), "evaluation.rule_ids"),
    }
    authorized = outcome in ("ALLOW", "MODIFY")
    if root.get("execution_authorized") is not authorized:
        raise StrictReceiptValidationError("execution_authorized disagrees with outcome")
    if outcome == "MODIFY":
        effective = _hash(
            action.get("effective_arguments_hash"), "action.effective_arguments_hash"
        )
        if effective == normalized_action["arguments_hash"]:
            raise StrictReceiptValidationError("MODIFY effective hash must differ")
        normalized_action["effective_arguments_hash"] = effective
    elif "effective_arguments_hash" in action:
        raise StrictReceiptValidationError("effective_arguments_hash requires MODIFY")

    suspension = None
    if record_type == "decision" and outcome in ("STEP_UP", "DEFER"):
        raw = _record(root.get("suspension"), "suspension")
        _exact(
            raw,
            {
                "suspension_id",
                "type",
                "status",
                "required_fields",
                "expires_at_ms",
                "approval_request_id",
                "approval_action_hash",
            },
            "suspension",
        )
        expected = "approval" if outcome == "STEP_UP" else "context"
        if raw.get("type") != expected:
            raise StrictReceiptValidationError(
                f"suspension.type must be {expected}"
            )
        if raw.get("status") != "pending":
            raise StrictReceiptValidationError("suspension.status must be pending")
        required_fields = _string_set(
            raw.get("required_fields"), "suspension.required_fields"
        )
        expires_at = _safe_integer(
            raw.get("expires_at_ms"), "suspension.expires_at_ms"
        )
        if expires_at < timestamp:
            raise StrictReceiptValidationError(
                "suspension.expires_at_ms precedes timestamp_ms"
            )
        suspension = {
            "suspension_id": _text(
                raw.get("suspension_id"), "suspension.suspension_id"
            ),
            "type": expected,
            "status": "pending",
            "required_fields": required_fields,
            "expires_at_ms": expires_at,
        }
        if outcome == "STEP_UP":
            if required_fields:
                raise StrictReceiptValidationError(
                    "STEP_UP required_fields must be empty"
                )
            suspension["approval_request_id"] = _text(
                raw.get("approval_request_id"),
                "suspension.approval_request_id",
            )
            suspension["approval_action_hash"] = _hash(
                raw.get("approval_action_hash"),
                "suspension.approval_action_hash",
            )
        else:
            if not required_fields:
                raise StrictReceiptValidationError(
                    "DEFER required_fields must be nonempty"
                )
            if "approval_request_id" in raw or "approval_action_hash" in raw:
                raise StrictReceiptValidationError(
                    "DEFER cannot carry approval fields"
                )
    elif "suspension" in root:
        raise StrictReceiptValidationError(
            "suspension requires a STEP_UP or DEFER decision"
        )

    resolution = None
    if record_type == "resolution":
        if outcome not in ("ALLOW", "DENY", "MODIFY"):
            raise StrictReceiptValidationError("resolution outcome must be final")
        raw = _record(root.get("resolution"), "resolution")
        _exact(
            raw,
            {
                "resolves_receipt_hash",
                "suspension_id",
                "method",
                "resolver_principal_id",
                "resolution_source_hash",
                "resolved_at_ms",
            },
            "resolution",
        )
        method = raw.get("method")
        if method not in (
            "approval_granted",
            "approval_denied",
            "context_supplied",
            "expired",
            "cancelled",
        ):
            raise StrictReceiptValidationError("invalid resolution.method")
        if method in ("approval_denied", "expired", "cancelled") and outcome != "DENY":
            raise StrictReceiptValidationError("resolution method requires DENY")
        resolved_at = _safe_integer(
            raw.get("resolved_at_ms"), "resolution.resolved_at_ms"
        )
        if resolved_at > timestamp:
            raise StrictReceiptValidationError(
                "resolution.resolved_at_ms exceeds timestamp_ms"
            )
        resolution = {
            "resolves_receipt_hash": _hash(
                raw.get("resolves_receipt_hash"),
                "resolution.resolves_receipt_hash",
            ),
            "suspension_id": _text(
                raw.get("suspension_id"), "resolution.suspension_id"
            ),
            "method": method,
            "resolver_principal_id": _text(
                raw.get("resolver_principal_id"),
                "resolution.resolver_principal_id",
            ),
            "resolution_source_hash": _hash(
                raw.get("resolution_source_hash"),
                "resolution.resolution_source_hash",
            ),
            "resolved_at_ms": resolved_at,
        }
    elif "resolution" in root:
        raise StrictReceiptValidationError("decision cannot carry resolution")

    body = {
        "schema": STRICT_RECEIPT_SCHEMA,
        "profile_version": STRICT_RECEIPT_PROFILE_VERSION,
        "record_type": record_type,
        "receipt_id": f"{session_id}:{sequence}",
        "session_id": session_id,
        "sequence": sequence,
        "timestamp_ms": timestamp,
        "clock_regression_clamped": root["clock_regression_clamped"],
        "previous_receipt_hash": previous,
        "sdk": {"language": sdk["language"], "version": _text(sdk.get("version"), "sdk.version")},
        "initiator": {
            "agent_id": _text(initiator.get("agent_id"), "initiator.agent_id"),
            "key_id": key_id,
        },
        "action": normalized_action,
        "context": normalized_context,
        "evaluation": normalized_evaluation,
        "execution_authorized": authorized,
    }
    if suspension is not None:
        body["suspension"] = suspension
    if resolution is not None:
        body["resolution"] = resolution
    return body


def canonicalize_strict_receipt_body(input_value: Any) -> str:
    return _canonical_json_for_hash(build_strict_receipt_body(input_value))


def strict_receipt_hash(input_value: Any) -> str:
    body = canonicalize_strict_receipt_body(input_value).encode("utf-8")
    preimage = STRICT_RECEIPT_BODY_DOMAIN + b"\x00" + _u64(len(body)) + body
    return hashlib.sha256(preimage).hexdigest()


def strict_receipt_signature_preimage(key_id: str, receipt_hash: str) -> bytes:
    if (
        not isinstance(key_id, str)
        or not key_id.startswith("sha256:")
        or len(key_id) != 71
        or any(char not in _HEX for char in key_id[7:])
    ):
        raise StrictReceiptValidationError("invalid strict key id")
    key = key_id.encode("utf-8")
    digest = bytes.fromhex(_hash(receipt_hash, "receipt_hash"))
    return STRICT_RECEIPT_SIGNATURE_DOMAIN + b"\x00" + _u64(len(key)) + key + digest


def sign_strict_receipt(
    input_value: Any, signer: Any, include_public_key: bool = False
) -> Dict[str, Any]:
    body = build_strict_receipt_body(input_value)
    key_id = strict_receipt_key_id(signer.raw_public_key)
    if body["initiator"]["key_id"] != key_id:
        raise StrictReceiptValidationError("signer does not match initiator.key_id")
    receipt_hash = strict_receipt_hash(body)
    preimage = strict_receipt_signature_preimage(key_id, receipt_hash)
    signature = signer.sign_bytes(preimage)
    if (
        not isinstance(signature, str)
        or len(signature) != 128
        or any(char not in _HEX for char in signature)
    ):
        raise StrictReceiptValidationError(
            "signer returned an invalid Ed25519 signature"
        )
    public_key_b64 = base64.b64encode(signer.raw_public_key).decode("ascii")
    if signer.public_key_b64 != public_key_b64:
        raise StrictReceiptValidationError(
            "signer public_key_b64 does not match raw_public_key"
        )
    from .policy_verify import _resolve_backend

    backend = _resolve_backend()
    try:
        signature_bytes = binascii.unhexlify(signature)
    except (binascii.Error, ValueError) as exc:
        raise StrictReceiptValidationError(
            "signer returned an invalid Ed25519 signature"
        ) from exc
    if backend is None or not backend(signer.raw_public_key, preimage, signature_bytes):
        raise StrictReceiptValidationError(
            "signer signature failed self-verification"
        )
    envelope = {
        "schema": STRICT_RECEIPT_ENVELOPE_SCHEMA,
        "body": body,
        "receipt_hash": receipt_hash,
        "signature": {
            "algorithm": "Ed25519",
            "key_id": key_id,
            "value": signature,
        },
    }
    if include_public_key:
        envelope["public_key_b64"] = public_key_b64
    return envelope

File: sdk-typescript/src/governance/strict-receipt.ts
import { createHash } from 'node:crypto';
import { canonicalJsonForHash } from '../policy/tool-pinning.js';
import type { DeviceSigner } from '../proxy/device-identity.js';
import { AARM_OUTCOMES, type AarmOutcome } from './aarm-outcome.js';

export const STRICT_RECEIPT_SCHEMA = 'obsvr-strict-receipt-v1' as const;
export const STRICT_RECEIPT_PROFILE_VERSION = '1.0' as const;
export const STRICT_RECEIPT_ENVELOPE_SCHEMA = 'obsvr-strict-receipt-envelope-v1' as const;
export const STRICT_RECEIPT_BODY_DOMAIN = 'obsvr-strict-receipt/body/1' as const;
export const STRICT_RECEIPT_SIGNATURE_DOMAIN = 'obsvr-strict-receipt/signature/1' as const;

const HEX64 = /^[0-9a-f]{64}$/;
const KEY_ID = /^sha256:[0-9a-f]{64}$/;
const OUTCOMES = new Set<string>(AARM_OUTCOMES);

export interface StrictReceiptBody {
  schema: typeof STRICT_RECEIPT_SCHEMA;
  profile_version: typeof STRICT_RECEIPT_PROFILE_VERSION;
  record_type: 'decision' | 'resolution';
  receipt_id: string;
  session_id: string;
  sequence: number;
  timestamp_ms: number;
  clock_regression_clamped: boolean;
  previous_receipt_hash: string | null;
  sdk: { language: 'typescript' | 'python'; version: string };
  initiator: { agent_id: string; key_id: string };
  action: {
    action_id: string;
    kind: string;
    name: string;
    arguments_hash: string;
    target?: string;
    effective_arguments_hash?: string;
  };
  context: {
    schema: 'obsvr-action-context-v1';
    context_hash: string;
    run_id: string;
    thread_id?: string;
  };
  evaluation: {
    input_hash: string;
    policy_hash: string;
    evaluator_hash: string;
    engine_version: string;
    policy_version: string;
    outcome: AarmOutcome;
    reason_code: string;
    rule_ids: string[];
  };
  execution_authorized: boolean;
  suspension?: {
    suspension_id: string;
    type: 'approval' | 'context';
    status: 'pending';
    required_fields: string[];
    expires_at_ms: number;
    approval_request_id?: string;
    approval_action_hash?: string;
  };
  resolution?: {
    resolves_receipt_hash: string;
    suspension_id: string;
    method: 'approval_granted' | 'approval_denied' | 'context_supplied' | 'expired' | 'cancelled';
    resolver_principal_id: string;
    resolution_source_hash: string;
    resolved_at_ms: number;
  };
}

export interface StrictReceiptEnvelope {
  schema: typeof STRICT_RECEIPT_ENVELOPE_SCHEMA;
  body: StrictReceiptBody;
  receipt_hash: string;
  signature: { algorithm: 'Ed25519'; key_id: string; value: string };
  public_key_b64?: string;
}

export class StrictReceiptValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'StrictReceiptValidationError';
  }
}

function record(value: unknown, field: string): Record<string, unknown> {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    throw new StrictReceiptValidationError(`${field} must be an object`);
  }
  return value as Record<string, unknown>;
}

function exact(value: Record<string, unknown>, allowed: readonly string[], field: string): void {
  const set = new Set(allowed);
  const unknown = Object.keys(value).filter((key) => !set.has(key)).sort();
  if (unknown.length) {
    throw new StrictReceiptValidationError(`${field} contains unsupported field: ${unknown[0]}`);
  }
}

function text(value: unknown, field: string): string {
  if (typeof value !== 'string' || value.trim().length === 0) {
    throw new StrictReceiptValidationError(`${field} must be a nonblank string`);
  }
  return value;
}

function hash(value: unknown, field: string): string {
  if (typeof value !== 'string' || !HEX64.test(value)) {
    throw new StrictReceiptValidationError(`${field} must be 64 lowercase hex characters`);
  }
  return value;
}

function safeInteger(value: unknown, field: string, positive = false): number {
  if (typeof value !== 'number' || !Number.isSafeInteger(value) || value < (positive ? 1 : 0)) {
    throw new StrictReceiptValidationError(`${field} must be a ${positive ? 'positive' : 'nonnegative'} safe integer`);
  }
  return value;
}

function optionalText(source: Record<string, unknown>, key: string, field: string): string | undefined {
  if (!Object.prototype.hasOwnProperty.call(source, key)) return undefined;
  return text(source[key], field);
}

function compareScalars(left: string, right: string): number {
  const a = Array.from(left, (char) => char.codePointAt(0) as number);
  const b = Array.from(right, (char) => char.codePointAt(0) as number);
  for (let i = 0; i < Math.min(a.length, b.length); i++) if (a[i] !== b[i]) return a[i] - b[i];
  return a.length - b.length;
}

function stringSet(value: unknown, field: string): string[] {
  if (!Array.isArray(value)) throw new StrictReceiptValidationError(`${field} must be an array`);
  const values = value.map((item, index) => text(item, `${field}[${index}]`));
  values.sort(compareScalars);
  return values.filter((item, index) => index === 0 || item !== values[index - 1]);
}

function u64(value: number): Buffer {
  const out = Buffer.alloc(8);
  out.writeBigUInt64BE(BigInt(value));
  return out;
}

export function strictReceiptKeyId(rawPublicKey: Uint8Array): string {
  const raw = Buffer.from(rawPublicKey);
  if (raw.length !== 32) throw new StrictReceiptValidationError('public key must be 32 raw bytes');
  return `sha256:${createHash('sha256').update(raw).digest('hex')}`;
}

export function buildStrictReceiptBody(input: StrictReceiptBody): StrictReceiptBody {
  const root = record(input, 'receipt body');
  exact(root, [
    'schema', 'profile_version', 'record_type', 'receipt_id', 'session_id', 'sequence',
    'timestamp_ms', 'clock_regression_clamped', 'previous_receipt_hash', 'sdk', 'initiator', 'action', 'context',
    'evaluation', 'execution_authorized', 'suspension', 'resolution',
  ], 'receipt body');
  if (root.schema !== STRICT_RECEIPT_SCHEMA) throw new StrictReceiptValidationError('invalid receipt schema');
  if (root.profile_version !== STRICT_RECEIPT_PROFILE_VERSION) throw new StrictReceiptValidationError('invalid receipt profile_version');
  if (root.record_type !== 'decision' && root.record_type !== 'resolution') throw new StrictReceiptValidationError('invalid record_type');
  const sessionId = text(root.session_id, 'session_id');
  const sequence = safeInteger(root.sequence, 'sequence', true);
  if (root.receipt_id !== `${sessionId}:${sequence}`) throw new StrictReceiptValidationError('receipt_id must equal session_id:sequence');
  const timestamp = safeInteger(root.timestamp_ms, 'timestamp_ms');
  if (typeof root.clock_regression_clamped !== 'boolean') {
    throw new StrictReceiptValidationError('clock_regression_clamped must be a boolean');
  }
  let previous: string | null;
  if (sequence === 1) {
    if (root.previous_receipt_hash !== null) throw new StrictReceiptValidationError('genesis previous_receipt_hash must be null');
    previous = null;
  } else {
    previous = hash(root.previous_receipt_hash, 'previous_receipt_hash');
  }

  const sdk = record(root.sdk, 'sdk');
  exact(sdk, ['language', 'version'], 'sdk');
  if (sdk.language !== 'typescript' && sdk.language !== 'python') throw new StrictReceiptValidationError('invalid sdk.language');
  const initiator = record(root.initiator, 'initiator');
  exact(initiator, ['agent_id', 'key_id'], 'initiator');
  const keyId = text(initiator.key_id, 'initiator.key_id');
  if (!KEY_ID.test(keyId)) throw new StrictReceiptValidationError('initiator.key_id must be sha256:<64 lowercase hex>');

  const action = record(root.action, 'action');
  exact(action, ['action_id', 'kind', 'name', 'arguments_hash', 'target', 'effective_arguments_hash'], 'action');
  const normalizedAction: StrictReceiptBody['action'] = {
    action_id: text(action.action_id, 'action.action_id'),
    kind: text(action.kind, 'action.kind'),
    name: text(action.name, 'action.name'),
    arguments_hash: hash(action.arguments_hash, 'action.arguments_hash'),
  };
  const target = optionalText(action, 'target', 'action.target');
  if (target !== undefined) normalizedAction.target = target;

  const context = record(root.context, 'context');
  exact(context, ['schema', 'context_hash', 'run_id', 'thread_id'], 'context');
  if (context.schema !== 'obsvr-action-context-v1') throw new StrictReceiptValidationError('invalid context.schema');
  const normalizedContext: StrictReceiptBody['context'] = {
    schema: 'obsvr-action-context-v1',
    context_hash: hash(context.context_hash, 'context.context_hash'),
    run_id: text(context.run_id, 'context.run_id'),
  };
  const threadId = optionalText(context, 'thread_id', 'context.thread_id');
  if (threadId !== undefined) normalizedContext.thread_id = threadId;

  const evaluation = record(root.evaluation, 'evaluation');
  exact(evaluation, [
    'input_hash', 'policy_hash', 'evaluator_hash', 'engine_version', 'policy_version',
    'outcome', 'reason_code', 'rule_ids',
  ], 'evaluation');
  if (typeof evaluation.outcome !== 'string' || !OUTCOMES.has(evaluation.outcome)) throw new StrictReceiptValidationError('invalid evaluation.outcome');
  const outcome = evaluation.outcome as AarmOutcome;
  const normalizedEvaluation: StrictReceiptBody['evaluation'] = {
    input_hash: hash(evaluation.input_hash, 'evaluation.input_hash'),
    policy_hash: hash(evaluation.policy_hash, 'evaluation.policy_hash'),
    evaluator_hash: hash(evaluation.evaluator_hash, 'evaluation.evaluator_hash'),
    engine_version: text(evaluation.engine_version, 'evaluation.engine_version'),
    policy_version: text(evaluation.policy_version, 'evaluation.policy_version'),
    outcome,
    reason_code: text(evaluation.reason_code, 'evaluation.reason_code'),
    rule_ids: stringSet(evaluation.rule_ids, 'evaluation.rule_ids'),
  };
  const authorized = outcome === 'ALLOW' || outcome === 'MODIFY';
  if (root.execution_authorized !== authorized) throw new StrictReceiptValidationError('execution_authorized disagrees with outcome');
  if (outcome === 'MODIFY') {
    const effective = hash(action.effective_arguments_hash, 'action.effective_arguments_hash');
    if (effective === normalizedAction.arguments_hash) throw new StrictReceiptValidationError('MODIFY effective hash must differ');
    normalizedAction.effective_arguments_hash = effective;
  } else if (Object.prototype.hasOwnProperty.call(action, 'effective_arguments_hash')) {
    throw new StrictReceiptValidationError('effective_arguments_hash requires MODIFY');
  }

  let suspension: StrictReceiptBody['suspension'];
  if (root.record_type === 'decision' && (outcome === 'STEP_UP' || outcome === 'DEFER')) {
    const raw = record(root.suspension, 'suspension');
    exact(raw, [
      'suspension_id', 'type', 'status', 'required_fields', 'expires_at_ms',
      'approval_request_id', 'approval_action_hash',
    ], 'suspension');
    const type = outcome === 'STEP_UP' ? 'approval' : 'context';
    if (raw.type !== type) throw new StrictReceiptValidationError(`suspension.type must be ${type}`);
    if (raw.status !== 'pending') throw new StrictReceiptValidationError('suspension.status must be pending');
    const requiredFields = stringSet(raw.required_fields, 'suspension.required_fields');
    const expiresAt = safeInteger(raw.expires_at_ms, 'suspension.expires_at_ms');
    if (expiresAt < timestamp) throw new StrictReceiptValidationError('suspension.expires_at_ms precedes timestamp_ms');
    suspension = {
      suspension_id: text(raw.suspension_id, 'suspension.suspension_id'),
      type,
      status: 'pending',
      required_fields: requiredFields,
      expires_at_ms: expiresAt,
    };
    if (outcome === 'STEP_UP') {
      if (requiredFields.length !== 0) throw new StrictReceiptValidationError('STEP_UP required_fields must be empty');
      suspension.approval_request_id = text(raw.approval_request_id, 'suspension.approval_request_id');
      suspension.approval_action_hash = hash(raw.approval_action_hash, 'suspension.approval_action_hash');
    } else {
      if (requiredFields.length === 0) throw new StrictReceiptValidationError('DEFER required_fields must be nonempty');
      if (Object.prototype.hasOwnProperty.call(raw, 'approval_request_id')
        || Object.prototype.hasOwnProperty.call(raw, 'approval_action_hash')) {
        throw new StrictReceiptValidationError('DEFER cannot carry approval fields');
      }
    }
  } else if (Object.prototype.hasOwnProperty.call(root, 'suspension')) {
    throw new StrictReceiptValidationError('suspension requires a STEP_UP or DEFER decision');
  }

  let resolution: StrictReceiptBody['resolution'];
  if (root.record_type === 'resolution') {
    if (!['ALLOW', 'DENY', 'MODIFY'].includes(outcome)) throw new StrictReceiptValidationError('resolution outcome must be final');
    const raw = record(root.resolution, 'resolution');
    exact(raw, [
      'resolves_receipt_hash', 'suspension_id', 'method', 'resolver_principal_id',
      'resolution_source_hash', 'resolved_at_ms',
    ], 'resolution');
    const method = raw.method;
    if (!['approval_granted', 'approval_denied', 'context_supplied', 'expired', 'cancelled'].includes(String(method))) {
      throw new StrictReceiptValidationError('invalid resolution.method');
    }
    if (['approval_denied', 'expired', 'cancelled'].includes(String(method)) && outcome !== 'DENY') {
      throw new StrictReceiptValidationError('resolution method requires DENY');
    }
    const resolvedAt = safeInteger(raw.resolved_at_ms, 'resolution.resolved_at_ms');
    if (resolvedAt > timestamp) throw new StrictReceiptValidationError('resolution.resolved_at_ms exceeds timestamp_ms');
    resolution = {
      resolves_receipt_hash: hash(raw.resolves_receipt_hash, 'resolution.resolves_receipt_hash'),
      suspension_id: text(raw.suspension_id, 'resolution.suspension_id'),
      method: method as NonNullable<StrictReceiptBody['resolution']>['method'],
      resolver_principal_id: text(raw.resolver_principal_id, 'resolution.resolver_principal_id'),
      resolution_source_hash: hash(raw.resolution_source_hash, 'resolution.resolution_source_hash'),
      resolved_at_ms: resolvedAt,
    };
  } else if (Object.prototype.hasOwnProperty.call(root, 'resolution')) {
    throw new StrictReceiptValidationError('decision cannot carry resolution');
  }

  const body: StrictReceiptBody = {
    schema: STRICT_RECEIPT_SCHEMA,
    profile_version: STRICT_RECEIPT_PROFILE_VERSION,
    record_type: root.record_type,
    receipt_id: `${sessionId}:${sequence}`,
    session_id: sessionId,
    sequence,
    timestamp_ms: timestamp,
    clock_regression_clamped: root.clock_regression_clamped,
    previous_receipt_hash: previous,
    sdk: { language: sdk.language, version: text(sdk.version, 'sdk.version') },
    initiator: { agent_id: text(initiator.agent_id, 'initiator.agent_id'), key_id: keyId },
    action: normalizedAction,
    context: normalizedContext,
    evaluation: normalizedEvaluation,
    execution_authorized: authorized,
  };
  if (suspension) body.suspension = suspension;
  if (resolution) body.resolution = resolution;
  return body;
}

export function canonicalizeStrictReceiptBody(input: StrictReceiptBody): string {
  return canonicalJsonForHash(buildStrictReceiptBody(input));
}

export function strictReceiptHash(input: StrictReceiptBody): string {
  const body = Buffer.from(canonicalizeStrictReceiptBody(input), 'utf8');
  return createHash('sha256').update(Buffer.concat([
    Buffer.from(STRICT_RECEIPT_BODY_DOMAIN, 'utf8'), Buffer.from([0]), u64(body.length), body,
  ])).digest('hex');
}

export function strictReceiptSignaturePreimage(keyId: string, receiptHash: string): Buffer {
  if (!KEY_ID.test(keyId)) throw new StrictReceiptValidationError('invalid strict key id');
  const key = Buffer.from(keyId, 'utf8');
  const digest = Buffer.from(hash(receiptHash, 'receipt_hash'), 'hex');
  return Buffer.concat([
    Buffer.from(STRICT_RECEIPT_SIGNATURE_DOMAIN, 'utf8'), Buffer.from([0]),
    u64(key.length), key, digest,
  ]);
}

export function signStrictReceipt(
  input: StrictReceiptBody,
  signer: DeviceSigner,
  includePublicKey = false,
): StrictReceiptEnvelope {
  const body = buildStrictReceiptBody(input);
  const keyId = strictReceiptKeyId(signer.rawPublicKey);
  if (body.initiator.key_id !== keyId) throw new StrictReceiptValidationError('signer does not match initiator.key_id');
  const receiptHash = strictReceiptHash(body);
  const envelope: StrictReceiptEnvelope = {
    schema: STRICT_RECEIPT_ENVELOPE_SCHEMA,
    body,
    receipt_hash: receiptHash,
    signature: {
      algorithm: 'Ed25519',
      key_id: keyId,
      value: signer.signBytes(strictReceiptSignaturePreimage(keyId, receiptHash)),
    },
  };
  if (includePublicKey) envelope.public_key_b64 = signer.publicKeyB64;
  return envelope;
}
